### PR TITLE
[receiver/prometheusremotewritereceiver] Fix labeled histogram exemplar attachment

### LINKED ISSUE #50546: [receiver/prometheusremotewrite] Exemplars are copied to every target and scope that shares a metric name
### Component(s)

receiver/prometheusremotewrite

### What happened?

`collectExemplars` keys exemplars by scope name, scope version, metric name, metric type and a hash of the data labels. `extractAttributes` drops `job`, `instance`, `__name__` and every `otel_scope_*` label before hashing, so none of them reach the key. Neither does the unit. Each of those tells one output metric from another, so when a request carries the same metric name for two of them, both get the exemplars.

Two cases, run against `fcf6c871bb`. In each one a single exemplar was sent.

Two targets that differ only by `job`, exemplar sent by `svc-a/one`:

```
exemplars per target: map[svc-a/one@host-a:1 svc-b/two@host-a:1]
```

Two scopes that differ only by `otel_scope_schema_url`, exemplar sent by `.../a`:

```
scope schemaURL="https://example.com/b" exemplars=1
scope schemaURL="https://example.com/a" exemplars=1
```

One arrived, two were emitted, and `X-Prometheus-Remote-Write-Exemplars-Written` reports 1.

This is not #50334. That one is the histogram lookup omitting `AttrsHash`, which is a field the key already has. This is about fields the key does not have at all, on both the collection and the attachment side.

### Collector version

main, currently fcf6c871bb

### Environment information

_No response_

### OpenTelemetry Collector configuration

_No response_

### Log output

_No response_

### Additional context

The `job` and `instance` case is a small fix, they can go straight into `exemplarKey`. The scope and unit case looks like it wants the identity the receiver already builds in `createMetricIdentity`, rather than more strings added one at a time, so I would rather do the two separately.

### PR DESCRIPTION
**Pull request by thc1006**
#### Description

Histogram exemplars are collected with a key that includes the data-label hash, but the histogram lookup omitted that field. As a result, exemplars on labeled histograms were not found. The counter path already includes it, since #48752.

The histogram path also copied matching exemplars into `DataPoints().At(0)`. When multiple series were translated into the same OTel metric, this could attach an exemplar to a datapoint produced for an earlier series.

This change:

- uses the same label-aware key for collection and histogram lookup;
- attaches matching exemplars once to the datapoint appended by the current histogram conversion;
- does not modify an existing datapoint when conversion drops the histogram;
- attaches on the counter path to the first datapoint that series appended, the way the histogram path does, and consumes the map entry there too, so two metric identities built from the same labels cannot both publish the same exemplar;
- stops hashing data point attributes altogether, for the key and for finding the datapoint, which leaves `pdatautil` unimported here and moved to the indirect block of `go.mod`, and builds the key only when the request carries exemplars.

The regression is reproducible when the histogram and exemplar are contained in the same PRW 2.0 `TimeSeries`. Measured on `main` with a histogram carrying one ordinary label, alongside its exemplar, in one `TimeSeries`:

| histogram | exemplars attached | Exemplars-Written |
| --- | --- | --- |
| no data labels | 1 | 1 |
| one data label | 0 | 1 |

This change does not add cross-request buffering or define exemplar-only wire semantics.

This addresses the histogram-association portion of #50334. Written-response semantics remain tracked in #50340.

##### What this does not change

`collectExemplars` runs over the whole request and keys by series, so exemplars belonging to one series are combined no matter which `TimeSeries` carried them, and the first matching data point takes all of them. This patch keeps that.

Measured with two `TimeSeries` sharing one label set, on `main`, on this branch, and on the counter path #48752 already merged: when only the second carries an exemplar, the first data point receives it; when both carry one, the first data point receives both. All three agree, so this is the receiver's existing series-level association rather than something this patch introduces.

The regression cases below keep the histogram and its exemplar in one `TimeSeries`, so they do not pin that inherited behaviour either way. If the reading that prometheus/docs#3085 is settling on calls for exemplars to stay with the `TimeSeries` that carried them, that is a change to the counter path as much as to this one.

##### Where this meets #50368

The two touch the same lines of `processHistogramTimeSeries` and do not merge cleanly. Resolving it here to see what it costs: `receiver.go` wants both the error returns #50368 gives the converters and the appended data point selected here, `translateV2` grows a fourth return that six call sites in these tests take, and `receiver_test.go` needs the two sets of new tests placed by hand rather than unioned.

Two things in this PR then stop earning their place. `TestDroppedHistogramExemplarDoesNotAttachToPreviousDataPoint` builds a histogram the converter used to drop without a word, which #50368 turns into a rejected request, so it fails. The `!appended` guard becomes unreachable, because both converters append before every path that returns nil. Whichever of the two lands second carries that work.

##### Which series an exemplar belongs to

The key `collectExemplars` built hashed the translated data point attributes, and `extractAttributes` drops `job`, `instance`, `__name__` and every `otel_scope_` label before hashing. Two targets sending one metric name shared a key, so on the earlier revision of this branch a labelled exemplar could land on the wrong target, and the counter path handed one received exemplar to both of them.

Metric type does not belong in it either. `QueueManager.Append` and `AppendExemplars` each read `t.seriesMetadata` at the moment they enqueue, so a sample queued before the metadata arrives carries `UNSPECIFIED` while an exemplar queued after it carries the real type. Keying on type would separate the two.

The key is the raw label set now. Prometheus identifies a series that way, so the metric name, the target and the scope labels are all in it, and nothing has to be listed by hand. Measured with two targets differing only by `job`, only by `instance`, only by `otel_scope_schema_url`, and only by a scope attribute: in each case the exemplar stays with the series that sent it, on both the counter and the histogram path. That is #50546.

#### Link to tracking issue

Part of #50334
Fixes #50546

#### Testing

Regression coverage, all with the histogram and its exemplar in the same `TimeSeries`:

| test | covers |
| --- | --- |
| `TestHistogramExemplarsAttachToLabelledSeries` | a labeled histogram finds its own exemplar |
| `TestHistogramExemplarsAttachToTheirOwnDataPoint` | two series in one metric, exponential and custom buckets |
| `TestHistogramExemplarsAreAttachedOnce` | two series under one key take the exemplar once |
| `TestDroppedHistogramExemplarDoesNotAttachToPreviousDataPoint` | a refused conversion leaves an earlier datapoint alone |
| `TestExemplarsAreClaimedByOneMetricOnly` | one exemplar, two identities from the same labels, six orderings |

The first three fail on `main`. The fourth passes there, because the missing label hash keeps the lookup from matching at all; it guards the path that adding the hash opens up, and it fails if the "did this conversion append a datapoint" check is removed.

`TestExemplarsAreClaimedByOneMetricOnly` is not a fail on `main` case, because the key there carries the metric type and so tells a COUNTER from an INFO apart by accident. It pins the consume step instead: with the `delete` on the counter path removed, five of its six orderings fail. The one that stays green is a histogram attaching before a counter, where the histogram path has already taken the entry.

`TestExemplarsStayWithTheirSeries` covers #50546 rather than the histogram path, on a counter and across two targets, with the four label sets measured above: differing only by `job`, only by `instance`, only by `otel_scope_schema_url`, and only by a scope attribute. All four subtests fail on `main`, each reporting the exemplar on two data points where one sent it.

#### Documentation

None. The behaviour being fixed is not described in the README.

#### Authorship

- [x] I, a human, wrote this pull request description myself.

### CODE REVIEW

File: receiver/prometheusremotewritereceiver/receiver.go [hunk 1]
```diff
@@ -475,17 +475,26 @@ func (prw *prometheusRemoteWriteReceiver) translateV2(_ context.Context, req *wr
 			addNumberDatapoints(metric.Sum().DataPoints(), ls, ts, &stats)
 			attrsHash := pdatautil.MapHash(extractAttributes(ls))
 			key := exemplarKey{
-				ScopeName:    si.Name,
-				ScopeVersion: si.Version,
-				MetricName:   metricName,
-				MetricType:   ts.Metadata.Type,
-				AttrsHash:    attrsHash,
+				Job:            ls.Get("job"),
```
Comment (krajorama) [inline receiver/prometheusremotewritereceiver/receiver.go]: The method `processHistogramTimeSeries`  has a way to calculate the correct hash I think, starting from `rm, hashedLabels = prw.getOrCreateRM(ls, otelMetrics, modifiedRM)` and ending in `metricIDHash := h.Sum64()`. Couldn't we just reuse that in `processHistogramTimeSeries` and also somehow outside it as well? Seems like now we'd have 3 implementations of the same thing, twice in the same function.

File: receiver/prometheusremotewritereceiver/receiver.go [hunk 2]
```diff
@@ -475,17 +475,26 @@ func (prw *prometheusRemoteWriteReceiver) translateV2(_ context.Context, req *wr
...
 			}
 			if ex, ok := exemplarMap[key.hash()]; ok && ex.Len() > 0 {
 				dataPoints := metric.Sum().DataPoints()
 				for i := 0; i < dataPoints.Len(); i++ {
 					if pdatautil.MapHash(dataPoints.At(i).Attributes()) == attrsHash {
 						ex.CopyTo(dataPoints.At(i).Exemplars())
+						// Reported as written only now, and consumed so that a repeated series
+						// cannot have the same exemplars counted or attached twice.
+						stats.Exemplars += ex.Len()
+						delete(exemplarMap, key.hash())
```
Comment (krajorama) [inline receiver/prometheusremotewritereceiver/receiver.go]: we're recalculating `key.hash()` here, let's move to a variable

File: receiver/prometheusremotewritereceiver/receiver.go [hunk 3]
```diff
@@ -617,30 +626,47 @@ func (prw *prometheusRemoteWriteReceiver) processHistogramTimeSeries(
...
+			MetricName:     metricName,
+			Unit:           unit,
+			MetricType:     ts.Metadata.Type,
+			AttrsHash:      pdatautil.MapHash(attrs),
 		}
 
 		if ex, ok := exemplarMap[key.hash()]; ok && ex.Len() > 0 {
 			ex.CopyTo(exemplarSlice)
+			stats.Exemplars += ex.Len()
+			delete(exemplarMap, key.hash())
```
Comment (krajorama) [inline receiver/prometheusremotewritereceiver/receiver.go]: we're recalculating key.hash() here, let's move to a variable

**Review by krajorama — COMMENTED** [review]
thank you for working on this, a few comments

File: receiver/prometheusremotewritereceiver/receiver.go (hunk 1 above)
Comment (thc1006) [reply receiver/prometheusremotewritereceiver/receiver.go]: Thanks. I consolidated the metric cache identity by adding HistogramType to metricIdentity and using metricIdentity.Hash() in processHistogramTimeSeries, so the inline histogram hash is gone.

I kept exemplarKey separate because collectExemplars builds it from the raw series labels before translation, while metricIdentity is built from the translated resource and scope. target_info can enrich that resource during translation, so using metricIdentity for the exemplar lookup would make the key depend on time series order.

The exemplar key hash is also now computed once on both the collection and attachment paths.

**Comment by dashpole** [comment]
> but Prometheus is still working out what senders guarantee here (https://github.com/prometheus/prometheus/issues/17857)

I'm a bit nervous about working around this, as it will be hard for us to remove this later without breaking people.  From the PRW 2.0 spec:

> "For every TimeSeries message:
At least one element in samples or in histograms MUST be provided. A TimeSeries MUST NOT include both samples and histograms."

So sending timeseries with only exemplars isn't compliant.

And from the proto definition it doesn't seem like we should be associating exemplars with timeseries they are not explicitly sent with:

```proto
// exemplars represents an optional set of exemplars attached to this series' samples.
repeated Exemplar exemplars = 4;
```

To me, it doesn't seem like we should be working around a bug in the Prometheus server by supporting this.  We likely won't be able to get rid of it, and IIUC, this imposes a performance penalty even for series with no associated exemplars. Would it be better to spend our effort fixing Prometheus? Or at least I would propose we update the PRW 2.0 spec/proto to allow this behavior if it is something we actually want to support.  @bwplotka for thoughts

**Comment by bwplotka** [comment]
We are actively discussing to lean on allowing this state as Prometheus exemplars are strictly per series and its hard to motivate per sample model switch at this time. So we will likely adjust the spec.

Fixing this without changing Prometheus model shifts performance penalty to client side, also adding penalty for ecosystems that relies on   pure per series association.

**Comment by thc1006** [comment]
Thanks, I agree the wire-format decision should stay separate from this receiver fix.

One clarification: the bug in #50334 does not need an exemplar-only `TimeSeries`. When a histogram and its exemplar are in the same `TimeSeries`, the exemplar is collected under a key that includes the data-label hash, but the histogram lookup omits it, so a histogram carrying any ordinary label never finds its own. The lookup also copies into `DataPoints().At(0)`, which is whichever series was translated first. #48752 fixed both of those on the counter path.

Measured on `main`, with the histogram and its exemplar in one `TimeSeries`:

| histogram | exemplars attached | Exemplars-Written |
| --- | --- | --- |
| no data labels | 1 | 1 |
| one data label | 0 | 1 |

I have narrowed this PR to that request-local bug. It is now 3 files, +232/-10:

- include the data-label hash in the histogram lookup key
- attach the exemplars once, to the datapoint the current conversion appended
- leave an existing datapoint alone when the conversion appends none
- skip the extra hashing when the request carries no exemplars

Out of this PR: the Written-header accounting, which belongs with the response semantics tracked in #50340, plus the target, scope and unit identity hardening, the trace and span ID length validation, and the metric cache cleanup. I have not added cross-request buffering, and this does not claim to solve prometheus/prometheus#17857.

File: receiver/prometheusremotewritereceiver/exemplars.go [hunk 4]
```diff
@@ -29,6 +27,7 @@ import (
 //   - instrumentation scope version
 //   - metric name
 //   - metric type
+//   - a hash of the remaining data labels
```
Comment (dashpole) [inline receiver/prometheusremotewritereceiver/exemplars.go]: Quick nit: this doc comment looks slightly out of date now that `makeExemplarKey` keys directly by the full label set (`ls.Bytes(nil)`) rather than hashing individual fields.

File: receiver/prometheusremotewritereceiver/receiver.go [hunk 5]
```diff
@@ -473,15 +473,9 @@ func (prw *prometheusRemoteWriteReceiver) translateV2(_ context.Context, req *wr
...
-			attrsHash := pdatautil.MapHash(extractAttributes(ls))
-			key := exemplarKey{
-				ScopeName:    si.Name,
-				ScopeVersion: si.Version,
-				MetricName:   metricName,
-				MetricType:   ts.Metadata.Type,
-				AttrsHash:    attrsHash,
-			}
-			if ex, ok := exemplarMap[key.hash()]; ok && ex.Len() > 0 {
+			key := makeExemplarKey(ls)
```
Comment (dashpole) [inline receiver/prometheusremotewritereceiver/receiver.go]: Should we guard this with `if len(exemplarMap) > 0` like in `processHistogramTimeSeries` so we skip allocating `ls.Bytes(nil)` when the request has no exemplars?

**Comment by thc1006** [comment]
My earlier comment here said the target, scope and unit identity hardening is out of this PR. That is not true anymore and I should have updated it. #50546 is folded in.

The reason is that the old key hashed the data point attributes, and `extractAttributes` removes `job`, `instance`, `__name__` and every `otel_scope_*` label before hashing. So two targets sending the same metric name got the same key, and one exemplar could be copied to both of them. The key is the whole label set now, which is how Prometheus identifies a series, so nothing has to be listed by hand. Type and unit are left out on purpose. `Append` and `AppendExemplars` each read `seriesMetadata` at the moment they enqueue, so a sample and an exemplar of one series can carry different types when the metadata arrives between them.

While checking that I found the two paths were not symmetric. The histogram path deletes the map entry after it attaches, but the counter path only copied it. So when the same labels build two metric identities, for example a COUNTER and an INFO, or the same type with two different units, both of them published the same exemplar. Both consume it now. `TestExemplarsAreClaimedByOneMetricOnly` covers six orderings. I took the `delete` out again to check the test really catches it: four of the six fail without it, and the two that stay green are the ones that were already safe.

One thing I did not change. The counter path still attaches to the first data point whose attributes match, not the one nearest the exemplar timestamp. When one `TimeSeries` carries several samples both choices are arbitrary, and moving it breaks an existing case that says it is testing exemplars sent disconnected from counters on purpose. That wants its own issue rather than this PR.

File: receiver/prometheusremotewritereceiver/receiver.go [hunk 6]
```diff
@@ -477,20 +481,19 @@ func (prw *prometheusRemoteWriteReceiver) translateV2(_ context.Context, req *wr
...
-				for i := 0; i < dataPoints.Len(); i++ {
-					if pdatautil.MapHash(dataPoints.At(i).Attributes()) == attrsHash {
-						ex.CopyTo(dataPoints.At(i).Exemplars())
-						break
+			// Consuming the entry stops a second metric identity built from the same labels,
+			// differing only by type or unit, from taking a copy of the same exemplars.
+			if len(exemplarMap) > 0 {
+				key := makeExemplarKey(ls)
+				if ex, ok := exemplarMap[key]; ok && ex.Len() > 0 {
+					attrsHash := pdatautil.MapHash(extractAttributes(ls))
```
Comment (dashpole) [inline receiver/prometheusremotewritereceiver/receiver.go]: Since `collectExemplars` runs over the whole request upfront, any series with samples already has its exemplars in `exemplarMap` when `addNumberDatapoints` appends to `metric.Sum().DataPoints()`. If you record `before := dps.Len()` before `addNumberDatapoints` and check `dps.Len() > before` (just like `processHistogramTimeSeries` does), you can attach directly to `dps.At(before).Exemplars()` in O(1) without scanning existing data points or calling `pdatautil.MapHash`, which also lets you drop the `pdatautil` import.

Comment (thc1006) [reply receiver/prometheusremotewritereceiver/receiver.go]: Done in 2f8467af. The scan and both `MapHash` calls are gone, and so is the `pdatautil` import.

Worth flagging that the index is load bearing, because I had tried something close to this earlier and picked the wrong one. `dps.At(before)` keeps the table case `counter metric with exemplar, multiple data points` passing, but `dps.At(dps.Len()-1)` fails it: that fixture puts two samples in one `TimeSeries` and expects the exemplars on the first of them. So which point gets them is pinned by a test rather than by how I read it.

It also made one of my own tests stricter, which I did not expect. `TestExemplarsAreClaimedByOneMetricOnly` has a `the same counter twice` case, and with the scan that case passed even with the `delete` taken out, because the scan found index 0 again and `CopyTo` overwrote the same slice. Attaching at `before` puts the second copy on its own data point, so removing the `delete` now fails 5 of the 6 orderings instead of 4.

File: receiver/prometheusremotewritereceiver/receiver_test.go [hunk 7]
```diff
@@ -191,6 +191,389 @@ func TestHandlePRWContentTypeNegotiation(t *testing.T) {
...
+			Exemplars:  exemplars,
+		}
+	}
+
+	metrics, _, err := prwReceiver.translateV2(t.Context(), &writev2.Request{
+		Symbols: exemplarSymbols,
+		Timeseries: []writev2.TimeSeries{
+			series(8, []int64{1, 1}, nil),
+			// The second bucket resolves to a negative count, so the converter drops it.
+			series(9, []int64{1, -5}, theExemplar()),
```
Comment (dashpole) [inline receiver/prometheusremotewritereceiver/receiver_test.go]: Since `validateExponentialHistogram` now rejects negative counts early at line 588, this exponential histogram is skipped before reaching `addExponentialHistogramDatapoint`, so this test passes even if the `!appended` check at line 692 is removed. Consider also testing an NHCB histogram with empty `CustomValues` (which returns early inside `addNHCBDatapoint`) so that the `!appended` guard is actually exercised.

**Comment by ArthurSens** [comment]
Just context for possible other reviewers, there's discussions happening at https://github.com/prometheus/prometheus/issues/17857 and https://github.com/prometheus/docs/pull/3085.

Let's put this PR on hold until upstream discussion settles on something

File: .chloggen/50334-prw-exemplar-association.yaml [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  all, and on the counter path one received exemplar could be written to both
+  targets. The key is the whole label set now, which is how Prometheus
+  identifies a series, so nothing that tells two series apart can be left out.
+  The histogram lookup also copied into the first data point of the metric,
+  which belongs to whichever series was translated first. Histogram exemplars go
+  to the data point the current conversion appended, and a histogram the
+  converter drops no longer reaches a data point built for an earlier series.
+  Both paths consume the entry once they have attached it, so a second metric
+  built from the same labels, differing only by type or unit, cannot publish a
+  second copy of what the sender wrote once.
```
Comment (ArthurSens) [inline .chloggen/50334-prw-exemplar-association.yaml]:
```suggestion
```

Looks irrelevant to the enduser 🤔

**Comment by thc1006** [comment]
Sounds good, that is fair. I made my case for this one back in August and it did not need repeating, so I have replaced what I first wrote here.

I will keep the branch rebased so it does not go stale, and will not push anything else to it while the upstream threads are open.

File: .chloggen/50334-prw-exemplar-association.yaml (hunk 8 above)
Comment (thc1006) [reply .chloggen/50334-prw-exemplar-association.yaml]: Done in 5211256743. That was the root cause rather than the change, and it belongs in the PR description.

If a shorter one would be more use than none, this is what I would put there, though only if you want it:

    Exemplars could be missing from histograms carrying an ordinary label, and one
    exemplar could be written to two targets sending the same metric name.

This is your own suggestion applied, so the hold stands.
